textattack/search_methods/improved_genetic_algorithm.py

```diff
@@ -0,0 +1,257 @@
+
+    Args:
+        max_pop_size (int): The maximum population size. Defaults to 20. 
+        max_iters (int): The maximum number of iterations to use. Defaults to 50.
+        max_replaced_times (int): The maximum number of times each word in the same position can be replaced at. Defaults to 5.
+        give_up_if_no_improvement (bool): If True, stop the search early if no candidate that improves the score is found.
+        max_crossover_retries (int): Maximum number of crossover retries if resulting child fails to pass the constraints.
+            Setting it to 0 means we immediately take one of the parents at random as the child.
+    """
+
+    def __init__(
+        self,
+        max_pop_size=20,
+        max_iters=50,
+        max_replaced_times=5,
+        give_up_if_no_improvement=False,
+        max_crossover_retries=20,
+    ):
+        self.max_iters = max_iters
+        self.max_pop_size = max_pop_size
+        self.max_replaced_times = max_replaced_times
+        self.give_up_if_no_improvement = give_up_if_no_improvement
+        self.max_crossover_retries = max_crossover_retries
+
+        # internal flag to indicate if search should end immediately
+        self._search_over = False
+
+    def _replace_at_index(self, pop_member, idx):
+        """
+        Select the best replacement for word at position (idx) 
+        in (pop_member) to maximize score.
+
+        Args:
+            pop_member: The population member being perturbed.
+            idx: The index at which to replace a word.
+
+        Returns:
+            Whether a replacement which increased the score was found.
+        """
+        transformations = self.get_transformations(
+            pop_member.attacked_text,
+            original_text=self.original_attacked_text,
+            indices_to_modify=[idx],
+        )
+        if not len(transformations):
+            return False
+        orig_result, self._search_over = self.get_goal_results(
+            [pop_member.attacked_text]
+        )
+        if self._search_over:
+            return False
+        new_x_results, self._search_over = self.get_goal_results(transformations)
+        new_x_scores = torch.Tensor([r.score for r in new_x_results])
+        new_x_scores = new_x_scores - orig_result[0].score
+        if len(new_x_scores) and new_x_scores.max() > 0:
+            max_idx = new_x_scores.argmax()
+            pop_member.attacked_text = transformations[max_idx]
+            pop_member.replaced_times[idx] = -1
+            pop_member.results = new_x_results[max_idx]
+            return True
+        return False
+
+    def _mutate(self, pop_member):
+        """
+        Replaces a word in pop_member that has not been modified. 
+        Args:
+            pop_member: The population member being perturbed.
+        """
+        x_len = pop_member.replaced_times.shape[0]
+        iterations = 0
+        while iterations < x_len and not self._search_over:
```

Comment:
One question. So I see that we try sampling `n` number of times where `n` is number of words. What happens if we for some reason end up unlucky and sample the same few indices that fail to produce transformations that improve the score? Should there be something that says "hey, don't sample this indices because it's bad". In the original GA method, this is done by explicitly defining the discrete probability distribution.